### PR TITLE
Fix to avoid stalling the process when ETW is doing a rundown

### PR DESCRIPTION
**Pull request by vancem**
This only matters when there are MANY JIT compiled methods, but Bing operates
in exactly this mode, and thus it stalls for several seconds  while rundown completes.

This fix does not fix the problem completely, but it makes it MUCH less likely, and is
a trivial, safe fix.  The problem is that as part of a GC, we do cleanup of any removed
JIT code.  To do this we take a JIT code manager lock, but this is also a lock that the
JIT code iterator takes and is used during ETW rundown.   Thus rundown blocks GCs.

Almost all the time, we DON'T have JIT code manager cleanup to do, so we just avoid
taking the lock in that case, and this makes the stall MUCH less likely.

### CODE REVIEW

File: src/vm/codeman.cpp [hunk 1]
```diff
@@ -3445,6 +3445,16 @@ void EEJitManager::CleanupCodeHeaps()
     }
     CONTRACTL_END;
 
+	// Quick out, don't even take the lock if we have not cleanup to do.
```
Comment (jkotas) [inline src/vm/codeman.cpp]: Could you please convert tabs to spaces?

**Comment by vancem** [comment]
I have moved the early out after the assert.

@dotnet-bot test this please

**Comment by vancem** [comment]
@dotnet-bot Windows_NT x64 Debug Build and Test

**Comment by vancem** [comment]
I have confirmed that with this change that in a common example managed code continues to run (allocate GC objects) while ETW is performing a rundown.   It has the effect desired (I was worried that there were other locks that also would end up becoming the critical issue after we cleared this one, but that is not the case).

**Comment by vancem** [comment]
This fix has also been ported to the .NET Desktop framework.  

See Bug 409381

https://devdiv.visualstudio.com/DefaultCollection/DevDiv/_workitems?id=409381&_a=edit

Resolved on 4/18/2017 (should be in 4.7.1 but I have not confirmed.

**Pull request merged**